### ISSUE TITLE
Feature idea: "Environment Variables" block

### ISSUE THREAD

**Issue by SelimRaboudi** [jupyter/jupyterlab#443]
Hi,

It is useful when doing interactive data analysis to quickly get the name/quick summary (type/ size) of in-memory variables.

I was wondering if having a block for displaying this kind of information could be an interesting feature for JupyterLab ? Such as the one in RStudio:

<img width="397" alt="screen shot 2016-07-18 at 15 55 52" src="https://cloud.githubusercontent.com/assets/2227272/16917136/a5f72562-4d00-11e6-80ae-3e275441e8a8.png">

**Comment by blink1073** [comment]
Hi @SelimRaboudi, this feature would be a most welcome contribution.

**Comment by sccolbert** [comment]
:+1: Would be a great project for a community developed extension.

**Comment by ellisonbg** [comment]
Great idea and +1 to someone from the community implementing this plugin. A design question: would you want to know the ENV vars for kernels or the main notebook server, or both?

**Comment by zertrin** [comment]
I'm afraid there is a misunderstanding between the title of this issue and the description.

OP seems to be asking more or less for the "Workspace Variables" block as can be seen in Matlab or similar programs, and not for the Environment variables which are a complete different thing.

Quoting him:
> get the name/quick summary (type/ size) of **in-memory variables**

I suggest changing the title of the issue to accurately reflect the description.

**Comment by SelimRaboudi** [comment]
@zertrin : indeed I've used the wrong vocabulary. Correcting the title.

**Comment by danielballan** [comment]
I heard a couple people at SciPy express interest in a feature like this. Could it be done with a thin wrapper around IPython's `who` magic? Is there a way to plug into cell execution such that `who` would be re-run and the widget updated every time a cell is executed?

**Comment by blink1073** [comment]
If not, we can add one ;).

**Comment by blink1073** [comment]
If you have the kernel, you can connect to `iopubMessage`, filter on `execute_result` messages, and take appropriate action.

**Comment by jasongrout** [comment]
(brainstorming here...)

Maybe we should have the concept of "kernel widgets" which are widgets that interact with a single kernel. You could open such a widget from a tab listing all running kernels, or maybe there would be a way to open the widgets from the kernel indication in a document. This variable explorer would be an example of a kernel widget. Likely a kernel widget could depend on the type of kernel, or it could be available for all kernels.

On the other hand, if this variable explorer was a kernel widget, what would we do if the notebook we launched it from started a different kernel? Would we want the variable explorer to automatically know to switch to the new kernel? I think that's probably what the user would expect, so the notebook that launched the kernel widget should keep a handle on it so it could change the kernel.

**Comment by svenefftinge** [comment]
Here again the notion of an "active" kernel could be useful, so the user doesn't need to select the kernel for such a widget. (mentioned in https://github.com/jupyter/jupyterlab/issues/450#issuecomment-233532869)

**Comment by danielballan** [comment]
I usually have more than one kernel running at a time, and I think I'd have trouble keeping track of which one was "active." I like @jasongrout's idea that kernel widgets like a "variable explorer" could be launched from a notebook and retain references in both directions. Other kernel widgets (like an HDF5 viewer) might close over a private kernel used only by that widget.

**Comment by aggFTW** [comment]
Love the idea of a kernel widget.

In our case, we'd love to develop a widget that knows the state of its kernel and displays information to the user based on it. In this case, the state is user variables, but in the case of a Spark kernel, it might be the connection string that the kernel is using to connect to Spark, as well as general information regarding this Spark instance.

If a user has multiple notebook tabs open, I believe good UX would be to have the widget change contents appropriately as the user selects multiple notebook tabs. One way of achieving this might be that kernel widgets are part of the notebooks JupyterLab panel, and can be minimized/maximized/docked/undocked in the context of the notebook panel.

**Comment by SynapticSage** [comment]
This feature would make JupyterLab a closer Pythonic analogue of Matlab's IDE

**Comment by OldGuyInTheClub** [comment]
Yes, please. With knobs on.  This is absolutely essential for scientific programming as is a visual debugger with breakpoints.

**Comment by Gerenuk** [comment]
The least it should show somehow is the keys and column names of dicts and tables. Looking up those identifier happens very often.

**Comment by vidartf** [comment]
Spyder IDE already has this for their IPython consoles (Spyder tries to be similar to Matlab IDE), and is licensed under MIT. Even if the language/framework used is different, it could serve as a source of inspiration.

**Comment by OldGuyInTheClub** [comment]
Spyder is ok but a friend recently showed me Rstudio and it was very impressive, very close to Matlab's UI and now has a notebook capability.  I don't think it fully supports other kernels though.  It is good enough that I am thinking about learning R.

**Comment by Juanlu001** [comment]
> This is absolutely essential for scientific programming as is a visual debugger with breakpoints.

Is there a separate issue/feature request for a visual debugger? I guess it should have to be developed as an extension. Perhaps this Python 3.7 addition will be of great help:

https://www.python.org/dev/peps/pep-0553/

**Comment by blink1073** [comment]
There isn't an issue for a debugger, please feel free to open one, @Juanlu001.

**Comment by Juanlu001** [comment]
Done! https://github.com/jupyterlab/jupyterlab/issues/3049

**Comment by sccolbert** [comment]
Not a gridpanel, but a datagrid (once we add the requisite features) (https://github.com/phosphorjs/phosphor/tree/master/packages/datagrid/src)

That gridpanel repo is old code that I still need to migrate to the 1.0 packages. It's a layout which arranges full blown widgets in a 2-d grid (as opposed to a boxpanel, which is 1d).

**Comment by azag0** [comment]
There is an [extension](https://github.com/ipython-contrib/jupyter_contrib_nbextensions/tree/master/src/jupyter_contrib_nbextensions/nbextensions/varInspector) for Jupyter Notebook that does pretty much this. I'm using it extensively. The code is actually quite simple. I don't know how difficult it is to adapt Jupyter Notebook extensions for Jupyterlab.

**Comment by jasongrout** [comment]
It should be pretty straightforward to adapt the code, and actually probably a lot of the window management code will not be needed.

**Comment by jasongrout** [comment]
(If anyone wants to work on this, please post here and/or in the gitter channel, and we can help you get started. You can also read the extension developer guide: http://jupyterlab.readthedocs.io/en/stable/user/extensions.html, http://jupyterlab.readthedocs.io/en/stable/developer/extension_dev.html, and peruse the xkcd example: http://jupyterlab.readthedocs.io/en/stable/developer/xkcd_extension_tutorial.html)

**Comment by sho-87** [comment]
@jasongrout I'm playing around with this. Nothing serious - mostly just a way for me to learn JS/TS and get familiar with jupyterlab. Doubt anything will come of it but we'll see...

Ive been looking through the docs to try to understand where things are in jupyterlab but its a bit sparse (understandable). How would I get a list of currently active (python) kernels, connect to one so I can execute some code, and receive a message back through iopub?

And are there any other tutorials on extension development that cover more advanced topics like these? Or is it mostly just sifting through the code base at this point and trying to piece things together from there?

**Comment by tbenschop** [comment]
Any news on this? Currently switching from R & RStudio to Python & JupyterLab and this is the only feature that I'm missing every minute.

**Comment by Aspire1Inspire2** [comment]
Yes, I agree, this is the only thing stopping me switching from spyder to jupyter.

**Comment by jasongrout** [comment]
I've opened #4409 about making a tutorial that covers starting a kernel from an extension.

@ian-r-rose - actually, I wonder if your help menu may be the simplest existing extension that uses a kernel.

**Comment by uptownfunk** [comment]
This would be so powerful guys, I think one of the few things left in having this be a serious contender to RStudio (that + support for RNotebooks if not already and RDashboards/Shiny dev)

**Comment by jakirkham** [comment]
> I've opened #4409 about making a tutorial that covers starting a kernel from an extension.

It sounds like one would want to access an existing (not start a new?) kernel for this. Though please feel free to correct me if I'm misunderstanding.

**Comment by kazuar** [comment]
Hi all,

Please see the short demo below.
It lists the existing variables and add / remove variables as you keep updating your notebook.

![variables_inspector](https://user-images.githubusercontent.com/422739/40275485-f5b5b1cc-5bbd-11e8-9ec8-194a9a978345.gif)

Still have to fix a few bugs before releasing something stable enough, I'll update more about my progress.

**Comment by jakirkham** [comment]
Neat! Thanks for sharing. What happens if you try a largeish array?

**Comment by kazuar** [comment]
@jakirkham it will probably print all of it in the output field. One more thing to fix 😄

**Comment by sho-87** [comment]
nice work @kazuar !

when I was playing around with this, before real life got in the way, I found the old varInspector extension to be a pretty nice starting point (https://github.com/ipython-contrib/jupyter_contrib_nbextensions/tree/master/src/jupyter_contrib_nbextensions/nbextensions/varInspector)

Maybe that will give you some ideas moving forward

Eager to see where you go with this!

**Comment by jasongrout** [comment]
> Still have to fix a few bugs before releasing something stable enough, I'll update more about my progress.

Great! Yes, please keep us up to date. There are a lot of people that would like to see a variable panel like this.

**Comment by kazuar** [comment]
Thanks @sho-87!
I did look at the Jupyter extension and got a few ideas from it. 
A few things are different mainly because JupyterLab's extensions are implemented with TypeScript but the general idea is still the same.

**Comment by kazuar** [comment]
@jakirkham Yes, putting `...` in case of reaching maximum length is exactly what I had in mind 😄

**Comment by kazuar** [comment]
@jasongrout Thanks! I've made the switch to JupyterLab a few weeks ago and I wanted to contribute after hearing a lot of great talks at the Jupyter Pop-up conference.

**Comment by kazuar** [comment]
Some more progress:

![variables_inspector2](https://user-images.githubusercontent.com/422739/40340074-3f3ebc06-5d4b-11e8-96b5-79cdfb459339.gif)

- Support switching between notebooks
- Support opening a new notebook
- Adding ellipsis `...` when value is too long to display
- A few bug fixes

Just a few more issues to take care of and I'll be able to release the extension and code.

**Comment by tbenschop** [comment]
Great stuff! Will this also be available for the editor? Or for a specific console?

**Comment by kazuar** [comment]
@tbenschop thanks! I only tested it on notebooks. I guess that for console it will require a few small changes because that would require using different interfaces.

**Comment by mlambelho** [comment]
Hey @kazuar, any news on the variable inspector? It looks extremely useful!

**Comment by mlambelho** [comment]
Hey @kazuar, any news on the variable inspector? It looks extremely useful! Btw, showing the size of the list or the dataframe could also be an additional cool feature.

**Comment by mlambelho** [comment]
Hey @kazuar, any news on the variable inspector? It looks extremely useful! Btw, showing the size of the list or the dataframe could also be an additional cool feature. 

Amazing job!

**Comment by kazuar** [comment]
Hi @mlambelho, thanks for your feedback and suggestion, that would be a great feature to add.
I've been traveling this week so not a lot of progress since last update. Hope to have more soon.

**Comment by onlyphantom** [comment]
Wow that’s a really neat extension. Also a much needed one to put Jupyter Lab in the same league as similar IDEs for the scientific computing audience - please keep us all updated!

**Comment by mkirch** [comment]
This is certainly on the right path and you are all doing great work. This is the one feature that would take Jupyter Lab to the next level. As suggested by a few other users above, I agree that it would be ideal to implement the R `data.frame` viewer for Pandas DataFrames and similar objects. Being able to visualize data structures makes working with data just _that_ much easier. For those unfamiliar, below is the way that this it works. The user can click on the `data.frame` object in the variable explorer, and it brings up a new tab with a sheet-style layout of the data contained in the `data.frame`. 

![out](https://user-images.githubusercontent.com/7461887/40992562-c76149ec-68bc-11e8-9046-a06fcb23c672.gif)

**Comment by rehans890** [comment]
@kazuar Hello, I am new to jupyerLab and stumbled upon this thread. Have you been able to finalize the extension?

**Comment by kazuar** [comment]
Hello @rehans890, still need to tie a few loose ends. Hoping to have some progress this weekend. Will update soon.

**Comment by codestar12** [comment]
@kazuar I'm super excited for this!! you rock!

**Comment by kazuar** [comment]
Thanks @codestar12! 
You probably rock way more than me 😄

**Comment by jakirkham** [comment]
Would you be open to sharing code/explaining how to install it? 😉 Seems like it has come a long way already and expect this will provide you a lot of useful feedback and contributions at this stage. 😄

**Comment by rajkar86** [comment]
Good work! If it's easy to implement sorting or filtering the inspector table by type, that would be a useful feature, too.

**Comment by lckr** [comment]
Hey all, I was missing a variable inspector from jlab for a while now. Today I hacked together a solution
and stumbled on this thread afterwards.
My extension supports python for now and is still in a very raw state. Nevertheless I pushed it to a [repo here](https://github.com/lckr/jupyterlab-variableInspector) .  
I will add some more documentation w/ screenshots tomorrow.
Feel free to take a look and add suggestions. You're very welcome to contribute.
I'm also very interested in your work @kazuar.  Maybe we can start a collaboration on this.

**Comment by kazuar** [comment]
@lckr Good job!
I've been busy every weekend for the last few weeks so it seems like I may have missed the mark on this one 😅.
From a general overview of you code it seems like both of our implementation kinda inspired by the same variable inspector ipwidget and jupyter's lab inspector.
If I'll ever get a free day over any weekend I'll push my code as well and I'll be more than happy to collaborate with you on anything you would like!

**Comment by Dodelo** [comment]
Hi @kazuar, your Demo looks awesome, is there any progress? Also, can we try it out as well? Thanks for your work!

**Comment by chappers** [comment]
@lckr has done a great job on his end: https://github.com/lckr/jupyterlab-variableInspector

As it works with both console and notebooks, and now with ability to inspect numpy and dataframes. There are still some teething issues which I have some suggestions on, but I suggest you guys check it out.

He has a demo in the readme and is shown here:

![demo](https://github.com/lckr/jupyterlab-variableInspector/raw/master/early_demo.gif)

**Comment by chappers** [comment]
Thought I'll give an update if people are still keen on what I'm working on (since, I don't know where is the interest level). I've been slowly adding support for things **I** use, being `spark` and `keras` + `tensorflow`

![image](https://user-images.githubusercontent.com/2498638/42431847-2238be48-838b-11e8-9d8e-15942e2ce8d4.png)

Which also supports "peeking" into it based on pd.get_options for spark.

![image](https://user-images.githubusercontent.com/2498638/42431881-4860d042-838b-11e8-8b3e-00b0ead2f77e.png)

**Comment by dclong** [comment]
How nicely does this work with the text editor? Can it be used to debug Python code in text editor rather than in notebooks?

**Comment by chappers** [comment]
@dclong it will work as expected when you use your text editor and attach a console to it.

**Comment by ajing** [comment]
I have following error when try to install the plugin

```
Errored, use --debug for full output:
ValueError:
"jupyterlab_variableinspector@0.1.0" is not compatible with the current JupyterLab
Conflicting Dependencies:
JupyterLab              Extension            Package
>=0.15.4-0 <0.16.0-0    >=0.16.0-0 <0.17.0-0 @jupyterlab/application
>=0.15.4-0 <0.16.0-0    >=0.16.3-0 <0.17.0-0 @jupyterlab/notebook
>=0.15.4-0 <0.16.0-0    >=0.16.3-0 <0.17.0-0 @jupyterlab/console
```

I am using jupyterlab 0.32.1

```
jupyterlab                0.32.1                   py36_0    conda-forge
```

**Comment by chappers** [comment]
@ajing probably better to raise it in the actual repo, but as suggested in your error, please try updating your jupyterlab to a version where the extensions are `>= 0.16.X`

**Comment by jakirkham** [comment]
Ran into [another variable inspector]( https://github.com/EtienneCmb/ipywksp ). So figured I'd share it here.

**Comment by koenlek** [comment]
@jakirkham, unfortunately it seems like the latest commit on that one is 3 years ago?

**Comment by edoardogiacomello** [comment]
How do you see the possibility to make it a TreeView like in PyCharm debugger?
![image](https://user-images.githubusercontent.com/9888197/54121950-04a12500-43fc-11e9-948a-f70873f7d394.png)

**Comment by zeluspudding** [comment]
I haven't used Jupyter notebooks for years because it doesn't have a variable inspector. I'd **love** to use Jupyter for the ability to make my research more shareable but not having an inspector is such a rough development experience.

I don't want this to sound like a complaint since this is open source software built thanks to untold thousands of volunteer hours. As an impassioned plea from someone who doesn't know how to build it himself - for the love of god - please make a native variable inspector! Lots of people want it:

- 75 upvotes on [this feature request](https://github.com/jupyterlab/jupyterlab/issues/443#issue-166101150)
- 32,000 views / 100 plus upvotes on [this stackoverflow request](https://stackoverflow.com/questions/37718907/variable-explorer-in-jupyter-notebook)

**Comment by Juanlu001** [comment]
@zeluspudding sorry but your comment indeed sounds like a complaint.

* we can expect dozens of people subscribed to this issue to receive status updates. making _impassionated pleas_ does not add anything useful.
* that stack overflow request describes several alternatives, even if they are not native
* even if you could not install jupyterlab-git, I don't think you opened an issue there describing your problems. or did you? https://github.com/jupyterlab/jupyterlab-git/issues/created_by/zeluspudding

I think the JupyterLab developers are well aware that this is a very requested feature (and Spyder has had it since, well forever). On the other hand, Jupyter has seen _massive_ adoption _despite not having it_, which is also a testament to its value as a project.

:peace_symbol:

**Comment by zeluspudding** [comment]
I am sorry my comment came across as such a complaint. Like I said, I've really wanted to use Jupyter for years - a testament that this group's work is incredibly valuable. 

I now notice that the [roadmap](https://github.com/jupyter/roadmap/blob/master/jupyterlab.md) lists this specifically. Can anyone comment as to an estimated time of release or whether this is still in the plans considering that there is an extension out there

**Comment by jasongrout** [comment]
I don't know of anyone working on this. However, we are having discussions about what a debugging protocol looks like, and that would probably include ways of getting information for a variable inspector. We don't have an estimated time for release for a debugging story (including a variable inspector) in the core extensions.

**Comment by vidartf** [comment]
That said, I'm sure people would be willing to help mentor anybody who says they are willing to commit the time to actually build such an extension.

**Comment by vidartf** [comment]
Considering that such an extension exists, I guess I should modify that to "anybody who's developing such an extension ".

**Comment by thbak** [comment]
I have installed the extension and it works perfectly - any chance it could support the Matlab kernel too?

**Comment by nielsenrechia** [comment]
Hello Guys,

I'm trying to use the variable inspector with R kernel. It shows to me the environment variables perfectly. My doubt is about opening data frames. It is possible with R kernel? In my case, It's only working with python or pyspark dataframes, but with R it is not.

Thanks,

**Comment by M00NSH0T** [comment]
it seems that the author of the variableinspector extension may not be actively maintaining it and so it's not compatible with JupyterLab3.  (see [here](https://github.com/lckr/jupyterlab-variableInspector/issues)). 

I'd love to see this functionality be rolled into the core JupyterLab functionality, as has been done with the debugger. As is, I'm sticking with JupyterLab2 for now just because I really love the variableinspector extension.

**Comment by jasongrout** [comment]
FYI, if I recall correctly, a variable inspector is on the roadmap for the debugger extension.

**Comment by jasongrout** [comment]
Updating that extension may be as simple as bumping the dependencies at https://github.com/lckr/jupyterlab-variableInspector/blob/50c522c613c6ce59fa19f88e4c7db11fa4b1dd56/package.json#L34-L40

**Comment by giswqs** [comment]
It is possible to return all defined variables as a list programmatically without using the magic command (`%who`) or the variable inspector? I need the list of variable names to be passed into a function. Thanks.

**Comment by jasongrout** [comment]
> Is it possible to return all defined variables as a list programmatically without using the magic command (%who) or the variable inspector? I need the list of variable names to be passed into a function. Thanks.

Sounds like you may want the Python [`globals()`](https://docs.python.org/3/library/functions.html#globals) function?

**Comment by giswqs** [comment]
> > Is it possible to return all defined variables as a list programmatically without using the magic command (%who) or the variable inspector? I need the list of variable names to be passed into a function. Thanks.
> 
> 
> 
> Sounds like you may want the Python [`globals()`](https://docs.python.org/3/library/functions.html#globals) function?

The list of various names defined in the notebook is to be passed into a function defined in a Python script. globals(), locals(), dirs() won't work in this case, as they only list the functions and variables defined in a Python script, not the actual ones defined in a notebook.

**Comment by jasongrout** [comment]
I meant to call `globals()` in the notebook and pass the resulting dictionary into your function imported from your python script.

**Comment by giswqs** [comment]
> I meant to call `globals()` in the notebook and pass the resulting dictionary into your function imported from your python script.

For example, the function below is defined in a module of a python package. When calling this function (from a *.py file) within Jupyter notebook, `globals()` actually gets all the variables/functions/classes defined in that module, not the ones defined in the notebook. 

```
def list_vars():

    result = []    
    for var in globals():
        if not var.startswith("_"):
            result.append(var)                
    return result
```

**Comment by jasongrout** [comment]
I think we are probably talking past each other, since it sounds like both of us understand how things work. My thought is that you could use `globals()` in a notebook cell (getting things defined in the notebook), and pass the result into a function.

**Comment by Atcold** [comment]
Are there any updates on this?

**Comment by anthonytec2** [comment]
From my understanding this was released recently: https://github.com/jupyterlab/jupyterlab/blob/f0790b3ffd187b7dc923156b8ae227855abecefb/packages/inspector/README.md. There is now a little bug on the side of the jupyterlab panel, where the variable inspector is.

**Comment by jakirkham** [comment]
Yeah thought this was done a while ago (unless I'm misunderstanding the ask)

https://blog.jupyter.org/a-visual-debugger-for-jupyter-914e61716559

**Comment by Atcold** [comment]
That requires using some niece kernel, where matplotlib and other libraries don't work.
I was just hoping one could glance at an interactive output of `%whos`.

**Comment by astrojuanlu** [comment]
The blog post is outdated: ipykernel, the mainstream kernel, supports the debugging protocol since 6.0.0, released in June 2021: https://github.com/ipython/ipykernel/releases/tag/6.0.0, https://jupyterlab.readthedocs.io/en/stable/user/debugger.html#requirements

![image](https://user-images.githubusercontent.com/316517/201500743-4b073929-4dab-4c65-8b2b-5a7e0af69efe.png)

**Comment by jakirkham** [comment]
Sounds like this issue should be closed then?

**Comment by Atcold** [comment]
I don't think so.
I just tried using the debugger for showing some `ndarray`s and `tensor`s, and it's basically unusable.
```python
import numpy
import torch
a = 3
b = 7
c = numpy.random.rand(4)
d = torch.rand(4)
```
![image](https://user-images.githubusercontent.com/2119355/202750455-3dbd63aa-b653-4eb6-82b5-c47b0ffc80c6.png)
![image](https://user-images.githubusercontent.com/2119355/202750606-a78ff4ad-cb8e-485b-b932-71105a07ce1f.png)
To be more explicit, for an `ndarray` and a `tensor` the shape should be reported. perhaps the `dtype`.

Using MATLAB (for comparison) we have

```matlab
a = 3
b = 7
c = rand([1, 4])
d = rand(4)
```
![image](https://user-images.githubusercontent.com/2119355/202754003-f505b703-2dc3-4287-8354-b88f7c8b2dfb.png)

**Comment by jakirkham** [comment]
That sounds like a new issue. Given this issue was about implementing the feature (as opposed to usability bugs with the feature)

**Comment by Atcold** [comment]
Fine with me.
Let's close this and I'll open the new one.

**Comment by krassowski** [comment]
Just to point out, since JupyterLab 3.3 you can hover over the list and click loop icon (or-right click to be offered an option to render the variable), as in:

![demo_render_variable](https://user-images.githubusercontent.com/8435071/120511259-89f0c580-c3ca-11eb-8572-5996fb0c6639.gif)

Currently it uses the default MIME type but we should be able to do better (like render a table for tabular pandas/numpy data even if currently we show just the default HTML preview).

@Atcold Did you open a new issue? We have a triage process which means that comments on old issues may not receive the same attention as new issues (and definitely we cannot tag a meta-issue for any specific release).

@jakirkham I personally would not consider this one closed as it required a debugger-enabled kernel.
